ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
 {
-  if (rc == MEMCACHED_SUCCESS)
-    return;
-
-  VALUE rb_error, boom;
-  const char * message;
-  if (rc > 0 && rc < MEMCACHED_ERROR_COUNT) {
-    rb_error = rb_eMemcachedErrors[rc];
-    if(Qnil == rb_error)
-      return;
-    message = memcached_strerror(NULL, rc);
-  } else {
-    rb_error = rb_eMemcachedError;
-    message = "Memcached returned type not handled";
-  }
-
-  boom = rb_exc_new2(rb_error, message);
-  rb_exc_raise(boom);
-}
-
-#define Wrap_return_t(rc) \
-{ \
-  handle_memcached_return(rc); \
-  if(rc == MEMCACHED_BUFFERED) \
-    return Qnil; \
-  else \
-    return rc == MEMCACHED_SUCCESS ? Qtrue : Qfalse; \
+	VALUE rb_err;
+	const char * message;
+
+	if (rc == MEMCACHED_SUCCESS)
+		return;
+
+	if (rc > 0 && rc < MEMCACHED_ERROR_COUNT) {
+		rb_err = rb_eMemcachedErrors[rc];
+		if (NIL_P(rb_err))
+			return;
+		message = memcached_strerror(NULL, rc);
+	} else {
+		rb_err = rb_eMemcachedError;
+		message = "Memcached returned type not handled";
+	}
+
+	rb_exc_raise(rb_exc_new2(rb_err, message));
 }
 
-memcached_ctx *
-get_ctx(VALUE obj)
-{
-  memcached_ctx* ctx;
-  Data_Get_Struct(obj, memcached_ctx, ctx);
-  return ctx;
-}
+#define rb_memcached_return(rc) { \
+		rb_memcached_error_check(rc); \
+		switch (rc) { \
+		case MEMCACHED_BUFFERED: return Qnil; \
+		case MEMCACHED_SUCCESS: return Qtrue; \
+		default: return Qfalse; }}
 
-static void
-free_memc(memcached_ctx * ctx)
-{
-  memcached_free(ctx->memc);
-}
 
 static VALUE
-allocate(VALUE klass)
+rb_memcached_alloc(VALUE klass)
 {
-  memcached_ctx * ctx;
-  VALUE rb_obj = Data_Make_Struct(klass, memcached_ctx, NULL, free_memc, ctx);
-
-  ctx->memc = memcached_create(NULL);
-  return rb_obj;
+	memcached_st *mc = memcached_create(NULL);
+	return Data_Wrap_Struct(klass, NULL, memcached_free, mc);
 }
 
 static VALUE
-rb_connection_initialize(VALUE self, VALUE rb_servers)
+rb_connection_new(VALUE klass, VALUE rb_servers)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  int i;
-  memcached_return_t rc;
-
-  for (i = 0; i < RARRAY_LEN(rb_servers); ++i) {
-    VALUE rb_server = rb_ary_entry(rb_servers, i);
-
-    VALUE rb_backend = rb_ary_entry(rb_server, 0);
-
-    if (id_tcp == SYM2ID(rb_backend)) {
-      VALUE rb_hostname = rb_ary_entry(rb_server, 1);
-      VALUE rb_port = rb_ary_entry(rb_server, 2);
-      // TODO: add weight
-      //VALUE rb_weight = rb_ary_entry(rb_server, 3);
-      rc = memcached_server_add (ctx->memc, StringValueCStr(rb_hostname), NUM2INT(rb_port));
-      handle_memcached_return(rc);
-    } else if (id_socket == SYM2ID(rb_backend)) {
-      VALUE rb_fd = rb_ary_entry(rb_server, 1);
-      rc = memcached_server_add_unix_socket (ctx->memc, StringValueCStr(rb_fd));
-      handle_memcached_return(rc);
-    }
-  }
-
-  // Verify keys by default
-  rc = memcached_behavior_set(ctx->memc, MEMCACHED_BEHAVIOR_VERIFY_KEY, true);
-  handle_memcached_return(rc);
-
-  return Qnil;
+	VALUE rb_mc = rb_memcached_alloc(klass);
```

Comment:
We could have created a `memcached_st` directly here, but note that we're allocating a Ruby object for it *and then* unwrapping it so we can perform operations. The reason is simple: if our initializer fails (raises), since the Ruby object has already been created, it won't leak. The GC will pick it up automatically.